cmake/jsoncpp.cmake: update to jsoncpp v1.8.4

aarlt opened the pull request:
Fixes #2829.

To get version 1.8.4 to work, we need to get rid of the usage of deprecated API. PR #3532 (libdevcore/JSON.h - new API) will remove the deprecated API calls.

Remember, that `scripts/travis-emscripten/build_emscripten.sh` is currently using the systems `cmake`, that is shipped with `trzeci/emscripten:sdk`, it is not using the one that is installed with `scripts/install_cmake.sh`.

The new `jsoncpp` (at least version 1.8.4) needs a minimum `cmake` version of 3.1, that is currently not provided by `trzeci/emscripten:sdk`, at least not with version `1.37.33-64bit`. So minor changes on the build scripts are needed.

chriseth commented:
The jsoncpp version we currently use also does not support move semantics, so this would also be a big benefit!

chriseth commented:
Please rebase and remove the merge commit, it is not possible to review it properly.

chriseth commented on libdevcore/JSON.h:
```diff
@@ -23,22 +23,31 @@
-	Json::FastWriter writer;
-	writer.omitEndingLineFeed();
-	return writer.write(_input);
+	std::stringstream stream;
```
Please combine this code with the one above.

aarlt replied:
Sure!

chriseth commented on libdevcore/JSON.h:
```diff
@@ -23,22 +23,31 @@
-	return writer.write(_input);
+	std::stringstream stream;
+	Json::StreamWriterBuilder builder;
+	builder.settings_["indentation"] = "";
```
The compact layout should not even have newlines. Does it  behave like that?

aarlt replied:
Yes, the behaviour is as expected. To be very sure, I will add a JSON Testsuite that will cover that - `test/libdevcore/JSON.cpp`.

chriseth commented on libsolc/libsolc.cpp:
```diff
@@ -229,12 +229,14 @@ string compile(StringMap const& _sources, bool _optimize, CStyleReadFileCallback
 	Json::Value input;
-	if (!reader.parse(_input, input, false))
+	std::stringstream inputStream(_input);
+	Json::CharReaderBuilder readerBuilder;
```
Can you write a helper function that does all this for us? It could even re-use a static variable of type `Json::CharReaderBuilder`.

aarlt replied:
Yeah sure!

chriseth reviewed (dismissed):
Please remove code repetitions.

aarlt commented:
I wasn't able to run the ipc specific tests with `eth`. I used the most recent version of `develop`. After running
- `eth/eth  --test -d /tmp/ipc` 
- `test/soltest -- --ipcpath /tmp/ipc/geth.ipc --no-smt`

```
➜  cmake-build-debug git:(strict-mode-jsoncpp-1.8.4) ✗ test/soltest -- --ipcpath /tmp/ipc/geth.ipc --no-smt
Running 1733 test cases...
/Users/alex/git/aarlt/solidity/test/RPCSession.cpp:332: fatal error: in "SolidityAuctionRegistrar/creation": critical check dev::jsonParseStrict(reply, &result, &errs) has failed
/Users/alex/git/aarlt/solidity/test/RPCSession.cpp:332: fatal error: in "SolidityAuctionRegistrar/reserve": critical check dev::jsonParseStrict(reply, &result, &errs) has failed
/Users/alex/git/aarlt/solidity/test/RPCSession.cpp:332: fatal error: in "SolidityAuctionRegistrar/double_reserve_long": critical check dev::jsonParseStrict(reply, &result, &errs) has failed
...
```
```
^[[1;37mcpp-ethereum, a C++ Ethereum client^[[0m
cpp-ethereum 1.3.0
  By cpp-ethereum contributors, (c) 2013-2018.
  See the README for contributors and credits.
Networking disabled. To start, use netstart or pass --bootstrap or a remote host.
JSONRPC Admin Session Key: GZeXowHW3aQ=
^[[91m ⚡    ^[[35m23:34:13.065^[[0m^[[30m|^[[34m^[[0m  Stop worker^[[0m 590 ms
^[[94m  ℹ  ^[[35m23:34:13.066^[[0m^[[30m|^[[34m^[[0m  Killing blockchain & extras database (WithExisting::Kill).
^[[91m ⚡    ^[[35m23:34:14.078^[[0m^[[30m|^[[34meth^[[0m  Worker stopping^[[0m 1013 ms
^[[101m^[[1;30m  ✘  ^[[35m23:34:14.328^[[0m^[[30m|^[[34meth^[[0m  Sender: dde147e35eedb9384833d66a487f1f246fb3406b^[[0m  Invalid Nonce: Require ^[[34m0^[[0m  Got ^[[34m3^[[0m
^[[101m^[[1;30m  ✘  ^[[35m23:34:14.328^[[0m^[[30m|^[[34meth^[[0m  Sender: dde147e35eedb9384833d66a487f1f246fb3406b^[[0m  Invalid Nonce: Require ^[[34m0^[[0m  Got ^[[34m3^[[0m
^[[101m^[[1;30m  ✘  ^[[35m23:34:14.328^[[0m^[[30m|^[[34meth^[[0m  Sender: dde147e35eedb9384833d66a487f1f246fb3406b^[[0m  Invalid Nonce: Require ^[[34m0^[[0m  Got ^[[34m3^[[0m
...
```
A lot of ipc related tests seem not to work. Sometimes `eth` just crashes. Am I doing something wrong? Should I file an issue?

However, because of that I couldn't run the ipc related tests. I was also not able to test my changes in `solfuzzer` - I didn't find out how it is used.

Additionally to this, it looks like that `jsoncpp` strict-mode still allow comments, even if the specific setting `allowComments` is set to `false`. 
```cpp
void CharReaderBuilder::strictMode(Json::Value* settings)
{
//! [CharReaderBuilderStrictMode]
  (*settings)["allowComments"] = false;
  (*settings)["strictRoot"] = true;
  (*settings)["allowDroppedNullPlaceholders"] = false;
  (*settings)["allowNumericKeys"] = false;
  (*settings)["allowSingleQuotes"] = false;
  (*settings)["stackLimit"] = 1000;
  (*settings)["failIfExtra"] = true;
  (*settings)["rejectDupKeys"] = true;
  (*settings)["allowSpecialFloats"] = false;
//! [CharReaderBuilderStrictMode]
}
```
It looks like a `jsoncpp` bug to me.

chriseth commented:
There are compilation errors on travis.

aarlt commented:
Oops.. its always good to use different compilers :)

aarlt commented:
On [Node.js:6 Compiler: gcc C++](https://travis-ci.org/ethereum/solidity/jobs/339178210)
```
CMake Error at CMakeLists.txt:3 (CMAKE_MINIMUM_REQUIRED):
  CMake 3.1 or higher is required.  You are running version 3.0.2
```
In `jsoncpp 1.8.4` they are using `CMAKE_MINIMUM_REQUIRED(VERSION 3.1)`, but here cmake seem to be `version 3.0.2`, that is contradicting to the output of `scripts/install_cmake.sh` some lines above. 
```
$ test "$TRAVIS_OS_NAME" != "linux" || (scripts/install_cmake.sh)
CMake 3.7.1 already installed in /home/travis/.local/bin
```

Is it possible that `version 3.0.2` was used all the time and the new `jsoncpp` is now the first thing that really needs `version 3.1` as minimum?

aarlt commented:
The [C++ travis build](https://travis-ci.org/ethereum/solidity/jobs/339178209) is not finding the make target for `deps/lib/libjsoncpp.a`. Somehow the lib seem to be installed in `/solidity/deps/lib64/libjsoncpp.a` - is `lib` vs. `lib64` the problem here?

chriseth commented:
Somehow it sometimes uses `lib` and somtimes `lib64`. I guess we need @chfast  to take a look at this...

chriseth commented:
Please also re-enable the fallthrough warning in cmake/jsoncpp.cmake

aarlt commented:
`jsoncpp` is compiled with `-Wall -Wconversion -Wshadow -Werror=conversion -Werror=sign-compare`, where `-Wimplicit-fallthrough` is only enabled, if its explicitly set, or if `-Wextra` is set - for [gnu compilers](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html). This is not the case for `jsoncpp`.

We can manually do `-Wimplicit-fallthrough`, but this seem to be only ok for `gnu` compilers - I tested it with `GCC 7.3.1`. If enabled by default, `clang` will still warn about the implicit fallthroughs, they are still there in `jsoncpp 1.8.4`. I tested `clang` versions `Apple LLVM version 9.0.0 (clang-900.0.39.2)` and fedora's `5.0.1`.

We shouldn't re-enable the `implicit-fallthrough`, because `jsoncpp` will still generate fallthrough warnings for `clang`. `jsoncpp` is not compiled with `-Wextra`, so we can get rid of the whole  implicit-fallthrough-block in `jsoncpp.cmake`.

However, do you mean "re-enable" in the sense of explicitly enabling `-Wimplicit-fallthrough`? 

I think that its safe to just remove the whole implicit-fallthrough-block in `jsoncpp.cmake`, it looks like that `-Wimplicit-fallthrough` and/or `-Wextra` is never set by the build system.

axic commented on libsolidity/interface/StandardCompiler.cpp:
```diff
@@ -236,7 +238,7 @@ Json::Value StandardCompiler::compileInternal(Json::Value const& _input)
 
 	Json::Value const& sources = _input["sources"];
-	if (!sources)
+	if (sources.empty())
```
Is this a bug fix?

aarlt replied:
Nope, they just removed `operator!()` in version `1.8.4`
```cpp
bool Value::operator!() const { return isNull(); }
```
Lol, now I was able to fully understand why I had problems using that operator. On my system where was somehow an old `jsoncpp` header used - will check the build system, how that was possible - and because they don't provided an implementation of `operator!()` I had linking problems. I just realised that they remove the `operator!()`, thats why I used `sources.empty()` here. Now I see that they also providing
```cpp
Value::operator bool() const { return ! isNull(); }
```
So everything is still fine. Wow, sorry for that!  The original code still works very nice with the new version `1.8.4`. I will remove that change.

aarlt replied:
... turned out it was a bug fix
```cpp
bool Value::empty() const {
  if (isNull() || isArray() || isObject())
    return size() == 0u;
  else
    return false;
}
```
In contrast to `Value::operator bool() `, `Value::empty()` is also checking whether an empty array, or an empty object was provided. I just added two new test-cases in `test/libsolidity/StandardCompiler.cpp`

axic replied:
Can you create a new PR which adds the test and removes the use of the `!` operator (without upgrading jsoncpp)?

aarlt replied:
Sure! See pull-request #3502.

chriseth commented:
We recently (6 months?) added some code to `cmake/jsoncpp.cmake` about fallthrough warnings. I though the new version might be fixed in that regard, so we could try removing that code.

aarlt commented:
Hmm.. looks like that `cmake/jsoncpp.cmake` also needs an ``if (EMSCRIPTEN)...``

aarlt commented:
```WARNING:root:generating system library: libcxx.a...
WARNING:root:                                      ok
warning: unresolved symbol: _ZN4Json5ValueC1EOS0_
warning: unresolved symbol: _ZN4Json5Value6appendEOS0_
...
missing function: _ZN4Json5Value6appendEOS0_
-1
...
abort(-1) at Error
    at jsStackTrace (/tmp/tmp.nehBzcMyfK/soljson.js:1:19718)
    at stackTrace (/tmp/tmp.nehBzcMyfK/soljson.js:1:19901)
    at abort (/tmp/tmp.nehBzcMyfK/soljson.js:19:10938)
    at __ZN4Json5Value6appendEOS0_ [Json::Value::append(Json::Value&&)] (/tmp/tmp.nehBzcMyfK/soljson.js:1:991699)
    at Array.Qbb (/tmp/tmp.nehBzcMyfK/soljson.js:11:418135)
    at Object.g7a [as dynCall_iii] (/tmp/tmp.nehBzcMyfK/soljson.js:11:398943)
    at invoke_iii (/tmp/tmp.nehBzcMyfK/soljson.js:1:1001029)
    at Array._j (/tmp/tmp.nehBzcMyfK/soljson.js:7:514811)
    at Object.y9a [as dynCall_viii] (/tmp/tmp.nehBzcMyfK/soljson.js:11:409096)
    at invoke_viii (/tmp/tmp.nehBzcMyfK/soljson.js:1:1002646)
If this abort() is unexpected, build with -s ASSERTIONS=1 which can give more information.
```
I just wondering.. "`__ZN4Json5Value6appendEOS0_`"  and "`_ZN4Json5Value6appendEOS0_`" are not the same symbols, the first one has two `_`.

aarlt commented:
I tried to fix the issue with the emscripten build. I saw that somehow a very old `cmake` was used, that prevented the build of `jsoncpp`, it required a minimum version `3.1`. After fixing that, during the build a missing function `_ZN4Json5Value6appendEOS0_` was shown, I did some debugging sessions.  After finding out that built-in  demangle support had some problems, I activated `DEMANGLE_SUPPORT=1` which didn't solved the problem - I thought that the buggy demangling caused somehow the optimiser to remove the function. However, I rewrote the `scripts/emscripten.sh` build, so it can be also used in OSX and updated to `trzeci/emscripten:sdk-tag-1.37.33-64bit`. The problem with the built-in demangling support seem to be fixed here. However, I saw there
```
RangeError: Maximum call stack size exceeded
    at JSON.stringify (<anonymous>)
    at Object.<anonymous> (/emsdk_portable/sdk/tools/js-optimizer.js:8389:16)
    at Module._compile (module.js:635:30)
    at Object.Module._extensions..js (module.js:646:10)
    at Module.load (module.js:554:32)
    at tryModuleLoad (module.js:497:12)
    at Function.Module._load (module.js:489:3)
    at Function.Module.runMain (module.js:676:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3
fs.js:646
  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
```

I tried to increment the `node_js` stack even more to 16MB, but I had no success. I disabled the JavaScript optimizer (`--js-opts 0`), everything worked, but the resulting `soljson.js` had a size of 75MB. After setting `EMCC_DEBUG=2` I was able to see the different optimizer steps:
```
DEBUG:root:emcc step "part of js opts" took 3.37 seconds
DEBUG:root:applying js optimization passes: asm aggressiveVariableElimination
chunkification: num funcs: 7053 actual num chunks: 9 chunk size range: 35726801 - 2026058
splitting up js optimization into 9 chunks, using 6 cores  (total: 58.66 MB)
.
.
.
.
.
/emsdk_portable/sdk/tools/eliminator/node_modules/uglify-js/lib/process.js:228
                                var ret = gen.apply(ast, ast.slice(1));
                                              ^

RangeError: Maximum call stack size exceeded
```
So somehow the `aggressiveVariableElimination` made problems, I disabled them so I was finally able to build a `soljson.js` (7.6MB). 

However, I tried now my findings on travis, but it looks like that there are still other problems.
```
warning: unresolved symbol: _ZNKSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE7compareEPKc
...
warning: unresolved symbol: _ZTINSt3__115basic_streambufIcNS_11char_traitsIcEEEE
warning: unresolved symbol: _ZTVNSt3__113basic_istreamIcNS_11char_traitsIcEEEE
```
Now it looks like that functions from the standard library are missing.  I have no idea why I don't see that locally. I ran the stuff also within a `trzeci/emscripten:sdk-tag-1.37.33-64bit` container.

Any ideas?

One question - because I want to push my version of `scripts/build_emscripten.sh`- does travis run everything in a container too? so if i create a named docker container inside travis, there will be no conflicts with other potential parallel builds using the same container-name?

axic commented:
> We recently (6 months?) added some code to cmake/jsoncpp.cmake about fallthrough warnings. I though the new version might be fixed in that regard, so we could try removing that code.

We added code to ignore fallthrough warnings, but later the fallthroughs were fixed and the ignore directive removed.

aarlt commented:
@chriseth do you know why the build was successful for `ci/circleci: build_emscripten`?

aarlt commented:
> > We recently (6 months?) added some code to cmake/jsoncpp.cmake about fallthrough warnings. I though the new version might be fixed in that regard, so we could try removing that code.
> 
> We added code to ignore fallthrough warnings, but later the fallthroughs were fixed and the ignore directive removed.
```cpp
# Disable implicit fallthrough warning in jsoncpp for gcc >= 7 until the upstream handles it properly
if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 7.0)
    set(JSONCCP_EXTRA_FLAGS -Wno-implicit-fallthrough)
else()
    set(JSONCCP_EXTRA_FLAGS "")
endif()
```
I removed the whole block, because from the compiler flags point of view it should never check for falltroughs. With `gnu` compilers they are only enabled, if explicitly set, or with `-Wextra` - that is not set in `JSONCCP_EXTRA_FLAGS`. And for `clang` it looks like that it need to be set explicitly with `-Wimplicit-fallthrough`.  If fallthroughs warnings are enabled, `clang` will still produce a lot of fallthrough warnings.

Probably the way how you built `jsoncpp` changed somehow - and you probably used `CMAKE_CXX_FLAGS` from `EthCompilerSettings.cmake` that sets `-Wextra` and `-Wimplicit-fallthrough`.

chriseth commented:
circle uses `trzeci/emscripten:sdk-tag-1.37.21-64bit` - try to see if you can use that for travis, too.

aarlt commented:
@chriseth interesting.. i tried 1.37.33-64 bit.. and there i had this optimizer bug with the `aggressiveVariableElimination`.. i will check that version now

axic commented:
~~#3491 does that but it failed~~ not sure where I got .19 from, probably an old version of the circleci PR.

aarlt commented:
Hmm.. is the `.travis.yml` in the repo not the real one used on travis?

axic commented:
Updated #3491 and that demangling issue happens there too even without the jsoncpp changes.

aarlt commented:
@chriseth `trzeci/emscripten:sdk-tag-1.37.21-64bit` failed also.. also some missing standard lib symbols.. it's really strange... there must be other differences..
```
(function (exports, require, module, __filename, __dirname) { var Module;if(!Module)Module=(typeof Module!=="undefined"?Module:null)||{};var moduleOverrides={};for(var key in Module){if(Module.hasOwnProperty(key)){moduleOverrides[key]=Module[key]}}var ENVIRONMENT_IS_WEB=false;var ENVIRONMENT_IS_WORKER=false;var ENVIRONMENT_IS_NODE=false;var ENVIRONMENT_IS_SHELL=false;if(Module["ENVIRONMENT"]){if(Module["ENVIRONMENT"]==="WEB"){ENVIRONMENT_IS_WEB=true}else if(Module["ENVIRONMENT"]==="WORKER"){ENVIRONMENT_IS_WORKER=true}else if(Module["ENVIRONMENT"]==="NODE"){ENVIRONMENT_IS_NODE=true}else if(Module["ENVIRONMENT"]==="SHELL"){ENVIRONMENT_IS_SHELL=true}else{throw new Error("The provided Module['ENVIRONMENT'] value is not valid. It must be one of: WEB|WORKER|NODE|SHELL.")}}else{ENVIRONMENT_IS_WEB=typeof window==="object";ENVIRONMENT_IS_WORKER=typeof importScripts==="function";ENVIRONMENT_IS_NODE=typeof process==="object"&&typeof require==="function"&&!ENVIRONMENT_IS_WEB&&!ENVIRONME
abort(-1) at Error
    at jsStackTrace (/tmp/tmp.wYO0uu72Zb/soljson.js:1:16688)
    at stackTrace (/tmp/tmp.wYO0uu72Zb/soljson.js:1:16859)
    at abort (/tmp/tmp.wYO0uu72Zb/soljson.js:19:11832)
    at __ZNSt3__16localeC1Ev (/tmp/tmp.wYO0uu72Zb/soljson.js:1:959617)
    at Array.KLa (/tmp/tmp.wYO0uu72Zb/soljson.js:5:633684)
    at Object.M5a [as dynCall_vi] (/tmp/tmp.wYO0uu72Zb/soljson.js:12:230974)
    at invoke_vi (/tmp/tmp.wYO0uu72Zb/soljson.js:1:974851)
    at Array.JLa (/tmp/tmp.wYO0uu72Zb/soljson.js:5:631740)
    at Object.Adb [as dynCall_iiiii] (/tmp/tmp.wYO0uu72Zb/soljson.js:12:275514)
    at invoke_iiiii (/tmp/tmp.wYO0uu72Zb/soljson.js:1:980976)
```
```
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE5uflowEv
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEC1ERKS5_jjRKS4_
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEaSERKS5_
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6insertENS_11__wrap_iterIPKcEEc
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE9push_backEc
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE9__grow_byEjjjjjj
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6appendEPKc
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE5imbueERKNS_6localeE
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE9underflowEv
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE4syncEv
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE5eraseEjj
warning: unresolved symbol: _ZNSt3__16localeC1Ev
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEC1ERKS5_
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE6xsgetnEPci
warning: unresolved symbol: _ZNSt3__18ios_base5clearEj
warning: unresolved symbol: _ZNSt3__113basic_istreamIcNS_11char_traitsIcEEED1Ev
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE9pbackfailEi
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEEC2Ev
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6__initEjc
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE8overflowEi
warning: unresolved symbol: _ZNKSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE7compareEPKc
warning: unresolved symbol: _ZNKSt3__121__basic_string_commonILb1EE20__throw_length_errorEv
warning: unresolved symbol: _ZNKSt3__120__vector_base_commonILb1EE20__throw_length_errorEv
warning: unresolved symbol: _ZNKSt3__120__vector_base_commonILb1EE20__throw_out_of_rangeEv
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEaSEc
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE6xsputnEPKci
warning: unresolved symbol: _ZNKSt3__16locale9use_facetERNS0_2idE
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEED1Ev
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6__initEPKcj
warning: unresolved symbol: _ZNSt3__18ios_base4initEPv
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6appendEjc
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6appendEPKcj
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE7reserveEj
warning: unresolved symbol: _ZNSt3__16localeD1Ev
warning: unresolved symbol: _ZNSt13runtime_errorC2ERKNSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEE
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEED2Ev
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6assignEPKc
warning: unresolved symbol: _ZNSt3__16localeC1ERKS0_
warning: unresolved symbol: _ZNSt3__19basic_iosIcNS_11char_traitsIcEEED2Ev
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6assignEjc
warning: unresolved symbol: _ZNSt3__16localeaSERKS0_
warning: unresolved symbol: _ZNKSt3__18ios_base6getlocEv
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6assignEPKcj
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE9showmanycEv
warning: unresolved symbol: _ZNSt3__113basic_istreamIcNS_11char_traitsIcEEErsERi
warning: unresolved symbol: _ZNSt3__15ctypeIcE2idE
warning: unresolved symbol: _ZNSt3__17collateIcE2idE
warning: unresolved symbol: _ZNSt3__18numpunctIcE2idE
warning: unresolved symbol: _ZTINSt3__115basic_streambufIcNS_11char_traitsIcEEEE
warning: unresolved symbol: _ZTVNSt3__113basic_istreamIcNS_11char_traitsIcEEEE
```

aarlt commented:
> Updated #3491 and that demangling issue happens there too even without the jsoncpp changes.

@axic good to know..

axic commented:
@aarlt as a next step would you be able to split out changes which introduce helpers for decoding/encoding without changing the jsoncpp version? I think we could merge that quickly.

After that this PR would only deal with a thin layer of jsoncpp version related changes.

aarlt commented:
@axic great!  I just noticed that the code is 100% compatible with jsoncpp 1.7.7!  the `Json::StreamWriterBuilder` & `Json::CharReaderBuilder` classes where already present in the old version! I will remove all the additional build-system changes - we can later just upgrade `jsoncpp.cmake` to version 1.8.4!

aarlt commented:
@axic I created a new PR #3532 (libdevcore/JSON.h - new API) that contains the major code changes - I will reduce the current PR #3467 to just contain the changes in `jsoncpp.cmake`.

So after merging #3532 the new `jsoncpp` version will hopefully just work - of course we need to fix the issues with the emscripten build. Additionally to that the emscripten build need to be changed abit, `scripts/travis-emscripten/build_emscripten.sh` is currently using the systems `cmake`, that is shipped with `trzeci/emscripten:sdk-tag-1.X.Y-64bit`, it is not using the one that is installed with `scripts/install_cmake.sh`. 

The new `jsoncpp` (at least version 1.8.4) needs a minimum cmake version of `3.1`, that is currently not provided by `trzeci/emscripten:sdk`, at least not within version `1.37.33-64bit`. But let's just wait until the `emscripten` linking problems are solved - probably `trzeci/emscripten:sdk` is then based on a more recent system.

aarlt committed:
CMake: Update jsoncpp to v1.8.4

chriseth commented:
Please also update `create_source_taball.sh`.

chfast commented on cmake/jsoncpp.cmake:
```diff
@@ -6,15 +6,17 @@ else()
-# Disable implicit fallthrough warning in jsoncpp for gcc >= 7 until the upstream handles it properly
-if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 7.0)
-    set(JSONCCP_EXTRA_FLAGS -Wno-implicit-fallthrough)
+get_property(LIB64 GLOBAL PROPERTY FIND_LIBRARY_USE_LIB64_PATHS)
```
I think there is easier way to do this: https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html.

```
include(GNUInstallDirs)

set(JSONCPP_LIBRARY "${prefix}/${CMAKE_INSTALL_LIBDIR}/...
```

aarlt replied:
Nice! Thats much better!

aarlt commented:
@chriseth Ok, I changed `scripts/create_source_tarball.sh` and `scripts/release_ppa.sh`

axic commented on scripts/release_ppa.sh:
```diff
@@ -76,7 +76,7 @@ mv solidity solc
 # Fetch jsoncpp dependency
 mkdir -p ./solc/deps/downloads/ 2>/dev/null || true
-wget -O ./solc/deps/downloads/jsoncpp-1.7.7.tar.gz https://github.com/open-source-parsers/jsoncpp/archive/1.7.7.tar.gz
+wget -O ./solc/deps/downloads/jsoncpp-1.8.4.tar.gz https://github.com/open-source-parsers/jsoncpp/archive/1.8.4.tar.gz
```
@chriseth why does the ppa script needs to download this separately? I thought cmake will do that on its own?

chfast replied:
Because you cannot download anything in PPA builds. So you have to prepare the sources yourself. If you put the archive in the place where CMake would download it, it will only check the hash and unzip it from there.

axic replied:
Not even with `cmake`? :)

chfast replied:
What I heard is the internet access is restricted only to official packages repo.

chriseth replied:
You have no internet access at all. You can declare dependencies on other packages, that's it, and I think this is a good thing: Building a package should be a deterministic, offline process.

axic commented:
@aarlt rebased, please pull before you push :)

axic commented on scripts/create_source_tarball.sh:
```diff
@@ -32,7 +32,7 @@ REPO_ROOT="$(dirname "$0")"/..
     # Add dependencies
     mkdir -p "$SOLDIR/deps/downloads/" 2>/dev/null || true
-    wget -O "$SOLDIR/deps/downloads/jsoncpp-1.7.7.tar.gz" https://github.com/open-source-parsers/jsoncpp/archive/1.7.7.tar.gz
+    wget -O "$SOLDIR/deps/downloads/jsoncpp-1.8.4.tar.gz" https://github.com/open-source-parsers/jsoncpp/archive/1.8.4.tar.gz
```
This is slightly unrelated to this PR, but I think this script should also check for the downloaded hash.

axic replied:
@chfast is it possible to run `cmake` to download deps only? :)

chfast replied:
It should be possible by `make jsoncpp-project-download`. If not, you have to "export" this ExternalProject step.

CMake will check the hash anyway, so no need to do it twice.

axic commented:
Now it fails with a different issue on Travis:

```
[  5%] Performing configure step for 'jsoncpp-project'
CMake Error at CMakeLists.txt:3 (CMAKE_MINIMUM_REQUIRED):
  CMake 3.1 or higher is required.  You are running version 3.0.2
```

(This was noted in #2829 that jsoncpp requires cmake 3.1)

aarlt commented:
@axic it always failed, it need to use the `cmake` that is installed with `scripts/install_cmake.sh` - its currently using the system version - that is really old. I solved that by just setting the `PATH` in `scripts/install_cmake.sh` and `source` it in the `build_emscripten.sh` - also `build.sh` have a problem here, but here its just ok, because its more recent.

aarlt commented:
@axic should the build scripts be changed during this PR?

chfast commented:
To fix empscriten build on Travis you can try to just upgrade the Emscripten docker image. But I usually had problems with this (weird build errors in other places).

axic commented:
@aarlt I guess it is worth a try.

@chfast that is currently being fixed/investigated in #3491

aarlt commented:
@chfast As far as I know, the system that is used in `trzeci/emscripten:sdk-tag-1.35.4-64bit` - and later version are quite old ubuntu's. Only older cmake versions provided there. 

@axic Ok, I will do that. But I guess it makes more sense to wait until you fixed the problems with `emscripten` and the old `jsoncpp v1.7.7` - after that I could fix the `cmake` version specific problem here. 
The strange thing is, that I solved the problem with `trzeci/emscripten:sdk-tag-1.37.33-64 bit` by remove that aggressive variable optimisation (that version had a bug in optimiser). Everything worked really nice within my docker using `trzeci/emscripten:sdk-tag-1.37.33-64 bit`, after pushing that to travis I also saw the unresolved symbols for the standard libs. I had really no explanation for that, probably something is going wrong on travis? The docker images should exactly behave identical, if same versions are used. At least I would expect that.

aarlt commented:
@axic cache cleaning sounds super good!

axic commented:
@aarlt it didn't fix it. We're not going ahead with the emscripten update at the moment. Can you try fixing cmake instead?

aarlt committed:
Install & use cmake provided by scripts/install_cmake.sh for emscripten

aarlt commented:
@axic ok, I did the changes.

axic commented:
@aarlt thanks! Can you please rebase the PR, not sure if there were any changes in the relevant files, but the build has been changed to be faster.

aarlt commented:
@axic PR rebased. Lets see whether its working ;)

chfast reviewed (dismissed):
Reviewed again. Looks good.

axic commented:
> What I heard is the internet access is restricted only to official packages repo.

That is the official jsoncpp repo.

axic commented:
Fails to link on travis:
```
[100%] Linking CXX executable soljson.js
WARNING:root:Assigning a non-existent settings attribute "NODEJS_CATCH_EXIT"
WARNING:root: - perhaps a typo in emcc's  -s X=Y  notation?
WARNING:root: - (see src/settings.js for valid values)
WARNING:root:generating system library: libcxx.a...
WARNING:root:                                      ok
WARNING:root:generating system library: libcxxabi.bc...
WARNING:root:                                          ok
error: unresolved symbol: _ZN4Json5ValueC1EOS0_
error: unresolved symbol: _ZN4Json5Value6appendEOS0_
```

aarlt commented:
Yep, the good old problem. I had it once running with `trzeci/emscripten:sdk-tag-1.37.33-64 bit`. But that failed on travis. I thought really that cleaning up the caches will fix the issue. The strange thing was that I was able to build the stuff within a `trzeci/emscripten:sdk-tag-1.37.33-64 bit` docker - on my machine everything was nice. On travis, with at this time exactly the same environment failed. Have you checked all potential caches? Is it possible to just do a clean build, that is not using any caching mechanisms? The big question is: why does it build on circleci?

axic commented:
Maybe jsoncpp uses some weird C++ construct that older version of emscripten cannot handle and decides those symbols aren't used. Not sure yet why the newer emscripten version is working on circleci, but not on travis :)

axic commented:
Demangling those two symbols:
- `Json::Value::Value(Json::Value&&)`
- `Json::Value::append(Json::Value&&)`

Both of these are defined under `#if JSON_HAS_RVALUE_REFERENCES`. My hunch is either we don't properly defined that (or the config is broken) or the old emscripten had issues with rvalue references (unlikely, since we use them).

chriseth commented:
I remember that the old jsoncpp did not have move semantics (i.e. rvalues), perhaps it is still linking to the old one but compiling with the new header?

chriseth commented:
Moved to 0.5.0

axic commented:
Need to revert https://github.com/ethereum/solidity/pull/3614 when working on this branch.

axic commented:
Rebased as I think this should be now working after #3888.

axic commented:
Sadly this is still not linking.

@chfast do you have any idea why it would still pick up and older version of the jsoncpp library?

That library line in cmake looks suspicious.

chfast commented:
@axic Can you show the logs?

axic commented:
It is the emscripten task of travis: https://travis-ci.org/ethereum/solidity/jobs/367535890

chfast commented on cmake/jsoncpp.cmake:
```diff
@@ -25,9 +19,9 @@ endif()
-    URL_HASH SHA256=087640ebcf7fbcfe8e2717a0b9528fff89c52fcf69fa2a18cc2b538008098f97
+    DOWNLOAD_NAME jsoncpp-1.8.4.tar.gz
+    URL https://github.com/open-source-parsers/jsoncpp/archive/1.8.4.tar.gz
+    URL_HASH SHA256=c49deac9e0933bcb7044f08516861a2d560988540b23de2ac1ad443b219afdb6
```
Can you try adding another CMAKE_ARGS: `-DCMAKE_CXX_STANDARD=11`?

axic replied:
Added.

axic committed:
CMake: enable C++11 for jsoncpp

axic commented:
Didn't help.

chfast commented:
So it's rather using new header but link the old lib.
Are you sure nothing is cached on Travis?

You can add `static_assert` in solidity code and check any of these: https://github.com/open-source-parsers/jsoncpp/blob/master/include/json/version.h#L7-L11.

axic commented:
My suspicion was the library part:

> @chfast do you have any idea why it would still pick up and older version of the jsoncpp library?
>
> That library line in cmake looks suspicious.

Cleared cache couple of times, but who knows.

axic commented on cmake/jsoncpp.cmake:
```diff
@@ -6,15 +6,9 @@ else()
+include(GNUInstallDirs)
 set(prefix "${CMAKE_BINARY_DIR}/deps")
-set(JSONCPP_LIBRARY "${prefix}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}jsoncpp${CMAKE_STATIC_LIBRARY_SUFFIX}")
+set(JSONCPP_LIBRARY "${prefix}/${CMAKE_INSTALL_LIBDIR}/${CMAKE_STATIC_LIBRARY_PREFIX}jsoncpp${CMAKE_STATIC_LIBRARY_SUFFIX}")
```
@chfast isn't this pointing to some system wide directory?

chfast replied:
`${CMAKE_INSTALL_LIBDIR}` should be just `"lib"`.

axic replied:
I mean together with `prefix`?

aarlt replied:
@chfast if `${CMAKE_INSTALL_LIBDIR}` would be just `lib`, we would have problems with `lib64` again, or am i wrong?

chfast commented on cmake/jsoncpp.cmake:
```diff
@@ -36,7 +31,8 @@ ExternalProject_Add(jsoncpp-project
                -DJSONCPP_WITH_PKGCONFIG_SUPPORT=OFF
-               -DCMAKE_CXX_FLAGS=${JSONCCP_EXTRA_FLAGS}
+	       -DCMAKE_CXX_FLAGS=${JSONCPP_EXTRA_FLAGS}
+               -DCMAKE_CXX_STANDARD=11
```
Try keeping `CMAKE_CXX_STANDARD=11` only, remove "extra flags".

axic replied:
You mean remove `EXTRA_FLAGS` above or remove the line adding c++11 to `EXTRA_FLAGS`?

axic replied:
Actually it seems the extra flags are needed as the warnings are back now. Also there's an excessive amount of warnings from cmake.

chfast replied:
Ok, so bring it back as it was but remove ` -DCMAKE_CXX_STANDARD=11`.

axic commented:
@aarlt this is now compiling :)

axic commented:
Without the `std=c++11` changes there were certain warnings:
```
-- Build files have been written to: /root/project/build/deps/src/jsoncpp-project-build
[  5%] Performing build step for 'jsoncpp-project'
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
Scanning dependencies of target jsoncpp_lib_static
[ 25%] Building CXX object src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/json_reader.cpp.o
In file included from /root/project/build/deps/src/jsoncpp-project/src/lib_json/json_reader.cpp:9:
In file included from /root/project/build/deps/src/jsoncpp-project/include/json/reader.h:11:
/root/project/build/deps/src/jsoncpp-project/include/json/value.h:404:3: warning: explicit conversion functions are a C++11 extension [-Wc++11-extensions]
  explicit operator bool() const;
  ^~~~~~~~
1 warning generated.
[ 50%] Building CXX object src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/json_value.cpp.o
In file included from /root/project/build/deps/src/jsoncpp-project/src/lib_json/json_value.cpp:8:
/root/project/build/deps/src/jsoncpp-project/include/json/value.h:404:3: warning: explicit conversion functions are a C++11 extension [-Wc++11-extensions]
  explicit operator bool() const;
  ^~~~~~~~
1 warning generated.
[ 75%] Building CXX object src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/json_writer.cpp.o
In file included from /root/project/build/deps/src/jsoncpp-project/src/lib_json/json_writer.cpp:7:
In file included from /root/project/build/deps/src/jsoncpp-project/include/json/writer.h:10:
/root/project/build/deps/src/jsoncpp-project/include/json/value.h:404:3: warning: explicit conversion functions are a C++11 extension [-Wc++11-extensions]
  explicit operator bool() const;
  ^~~~~~~~
1 warning generated.
[100%] Linking CXX static library libjsoncpp.a
```

chfast commented on cmake/jsoncpp.cmake:
```diff
@@ -36,7 +31,8 @@ ExternalProject_Add(jsoncpp-project
                -DJSONCPP_WITH_TESTS=OFF
                -DJSONCPP_WITH_PKGCONFIG_SUPPORT=OFF
-               -DCMAKE_CXX_FLAGS=${JSONCCP_EXTRA_FLAGS}
+	       -DCMAKE_CXX_FLAGS=${JSONCPP_EXTRA_FLAGS}
```
Remove this line.

chfast commented on scripts/travis-emscripten/build_emscripten.sh:
```diff
@@ -92,7 +102,7 @@ cmake \
   -DTESTS=0 \
   ..
-make -j 4
+make VERBOSE=1 -j 4
```
After successful build, this should be reverted.

chfast commented on .travis.yml:
```diff
@@ -161,7 +161,7 @@ install:
     # Disable tests unless run on the release branch, on tags or with daily cron
     #- if [ "$TRAVIS_BRANCH" != release -a -z "$TRAVIS_TAG" -a "$TRAVIS_EVENT_TYPE" != cron ]; then SOLC_TESTS=Off; fi
-    - SOLC_TESTS=Off
+    #- SOLC_TESTS=Off
```
After successful build, this should be probably reverted.

aarlt commented:
@axic yep its compiling :) i wrote that on gitter - but looks like u missed it :)

chfast commented on cmake/jsoncpp.cmake:
```diff
@@ -6,16 +6,11 @@ else()
-set(JSONCPP_LIBRARY "${prefix}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}jsoncpp${CMAKE_STATIC_LIBRARY_SUFFIX}")
+set(JSONCPP_LIBRARY "${prefix}/${CMAKE_INSTALL_LIBDIR}/${CMAKE_STATIC_LIBRARY_PREFIX}jsoncpp${CMAKE_STATIC_LIBRARY_SUFFIX}")
 set(JSONCPP_INCLUDE_DIR "${prefix}/include")
+set(JSONCPP_EXTRA_FLAGS "-std=c++11")
```
In the end it must be only for GCC / clang. You can wrap it with `if(NOT MSVC)`.

axic replied:
Are you sure? Appveyor seems to be building fine.

axic replied:
So Appveyor works with this, do you think it is still needed?

chfast replied:
Maybe it works because of this flag: https://msdn.microsoft.com/en-us/library/mt490614.aspx.
Let's leave it if AppVeyor is ok.

axic commented:
For some reason now Travis is emotional again:
```
Testing overwriting files
Testing soljson via the fuzzer...
Commandline tests successful.
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

axic commented:
@chriseth can you have a look at this?

axic commented:
@chriseth are you ok with this? Travis emscripten works. I'll remove the last commit if you're happy with this version before merging.

chriseth commented on scripts/install_cmake.sh:
```diff
@@ -18,6 +18,8 @@ esac
 
 BIN=$PREFIX/bin
 
+PATH=$PREFIX/bin:$PATH
```
Does this have any effect after the script has run?
